DOC: optimize: add `optimize.linear_sum_assignment` tutorial. (#13252)
scipy/scipy: doc/source/tutorial/optimize.rst

    success: True
          x: array([ 9.41025641,  5.17948718, -0.25641026,  1.64102564])  # may vary

Assignment problems
-------------------

Linear sum assignment problem example
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Consider the problem of selecting students for a swimming medley relay team.
We have a table showing times for each swimming style of five students:

==========  ===========  ============  ===========  ===============================
 Student    backstroke   breaststroke  butterfly    freestyle
==========  ===========  ============  ===========  ===============================
 A          43.5           47.1         48.4        38.2
 B          45.5           42.1         49.6        36.8
 C          43.4           39.1         42.1        43.2
 D          46.5           44.1         44.5        41.2
 E          46.3           47.8         50.4        37.2
==========  ===========  ============  ===========  ===============================

> We need to choose the best student for each of the four swimming styles.

Reviewer: ```suggestion
We need to choose a student for each of the four swimming styles such that 
the total relay time is minimized.
```

This is a typical linear sum assignment problem. We can use :func:`linear_sum_assignment` to solve it.

The linear sum assignment problem is one of the most famous combinatorial optimization problems.
Given a "cost matrix" :math:`C`, the problem is to choose

- exactly one element from each row 
- without choosing more than one element from any column 

> - such that the sum of the chosen elements is minimized (or maximized).

Reviewer: ```suggestion
- such that the sum of the chosen elements is minimized
```

In other words, we need to assign each row to one column such that the sum of 

> the corresponding entries is minimized (or maximized).

Reviewer: ```suggestion
the corresponding entries is minimized.
```

> Formally, let X be a boolean matrix where :math:`X[i,j] = 1` iff row i is assigned to column j.

Reviewer: ```suggestion
Formally, let :math:`X` be a boolean matrix where :math:`X[i,j] = 1` iff row i is assigned to column j.
```

Then the optimal assignment has cost

.. math::

    \min \sum_i \sum_j C_{i,j} X_{i,j}

> First of all, we need to setup a cost matrix.

Reviewer: ```suggestion
First of all, we need to set up a cost matrix.
```

> First of all, we need to setup a cost matrix.
> :func:`linear_sum_assignment` can assign each row to the best column.
> In this example, we want to assign each swimming style to a student.

Reviewer: ```suggestion
The first step is to define the cost matrix.
In this example, we want to assign each swimming style to a student.
:func:`linear_sum_assignment` is able to assign each row of a cost matrix to a column.
```

    success: True
          x: array([ 9.41025641,  5.17948718, -0.25641026,  1.64102564])  # may vary

> Linear sum assignment problem example

Reviewer: This should be put in its own section, "Assignment problems".
http://scipy.github.io/devdocs/optimize.html#assignment-problems

^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Consider planning a swimming medley relay team assignment problem.

> We have a table showing the individual medley times of five students:

Author: I'm not familiar with it too.. But I think you are right, the "medley time" is the sum of times of each style. So, I fixed it.

==========  ===========  ============  ===========  ===============================
 Student    backstroke   breaststroke  butterfly    freestyle
==========  ===========  ============  ===========  ===============================
 A          43.5           47.1         48.4        38.2
 B          45.5           42.1         49.6        36.8
 C          43.4           39.1         42.1        43.2
 D          46.5           44.1         44.5        41.2
 E          46.3           47.8         50.4        37.2

> ==========  ===========  ============  ===========  ===============================

Reviewer: If this data is taken from somewhere, the source would need to be acknowledged.

Author: This data was made up by me.

We need to choose best four students for each swimming style.
This is a typical linear sum assignment problem, we can use :func:`linear_sum_assignment` to solve it.

> First of all, we need to setup a cost matrix.

Author: OK. I tried to added it. PTAL.

:func:`linear_sum_assignment` can assign each row to the best column.

> We can setup the cost matrix based on the upper table:

Reviewer: Please explain why we need to transpose the table.

Author: I added. PTAL.

::

    >>> import numpy as np
    >>> cost = np.array([[43.5, 45.5, 43.4, 46.5, 46.3],
    ...                  [47.1, 42.1, 39.1, 44.1, 47.8],
    ...                  [48.4, 49.6, 42.1, 44.5, 50.4],
    ...                  [38.2, 36.8, 43.2, 41.2, 37.2]])

We can solve the assignment problem with :func:`linear_sum_assignment`:

::

    >>> from scipy.optimize import linear_sum_assignment
    >>> row_ind, col_ind = linear_sum_assignment(cost)
    >>> col_ind
    array([0, 2, 3, 1])

> The optimal assignment is:

Author: I added. PTAL.

::

    >>> dict(zip(["backstroke", "breaststroke", "butterfly", "freestyle"],
    ...          np.array(["A", "B", "C", "D", "E"])[col_ind]))

>     {'backstroke': 'A', 'breaststroke': 'C', 'butterfly': 'D', 'freestyle': 'B'}

Reviewer: To make it very obvious what's going on, I'd suggest breaking this into something like:
```python3
styles = np.array(["backstroke", "breaststroke", "butterfly", "freestyle"])[row_ind]
students = np.array(["A", "B", "C", "D", "E"])[col_ind]
dict(zip(styles, students))
```

Author: Thanks. I agree that is better. I fixed it.

The optimal times for each swimming style and the total time are:

::

    >>> cost[row_ind, col_ind]

>     array([43.5, 39.1, 44.5, 36.8])

Reviewer: My opinion (take it or leave it):
For selecting a team, I think I'd want to know which student is doing which style (as you have above) and what the total time is (below). The time for each style seems less important than those things, and calculating the total time below repeats most of this code, so consider taking these lines out.

Author: I agree. I removed it.

    >>> cost[row_ind, col_ind].sum()
    163.89999999999998

> 

Reviewer: I think that it's worth noting that this is not the same as the sum of the minimum times for each event:
```python3
np.min(cost, axis=1).sum()
```
and interpreting why (I haven't looked, but presumably one of the students was the best at more than one event.)

Author: Good idea. I added.

 E          46.3           47.8         50.4        37.2
==========  ===========  ============  ===========  ===============================

We need to choose the best student for each of the four swimming styles.
This is a typical linear sum assignment problem. We can use :func:`linear_sum_assignment` to solve it.

The linear sum assignment problem is one of the most famous combinatorial optimization problems.

> When we are given a matrix :math:`C` known as a "cost matrix" and we want to match each row to a different column,
> we need to find a combinations of a row and a column which the sum of the corresponding entries is minimized
> (or maximized).

Reviewer: ```suggestion
Given an :math:`m \times n` "cost matrix" :math:`C`, the problem is to choose

- exactly one element from each row 
- without choosing more than one element from any column 
- such that the sum of the chosen elements is minimized (or maximized).

In other words, we need to assign one column to each row such that the sum of the corresponding entries is minimized (or maximized).
```

Formally, let X be a boolean matrix where :math:`X[i,j] = 1` iff row i is assigned to column j.

> Then the optimal assignment has cost
> 
> .. math::
> 
>     \min \sum_i \sum_j C_{i,j} X_{i,j}

Reviewer: ```suggestion
Then the problem is

.. math::

   \min_X \sum_i \sum_j  & C_{i,j} X_{i,j} \\
   \text{s.t.} \quad \quad & \sum_i X_{i,j} = 1 \\ 
   & \sum_j X_{i,j} \leq 1
```
Or that's the best way I can figure out to write it in a way the Sphinx should interpret correctly.

First of all, we need to setup a cost matrix.
:func:`linear_sum_assignment` can assign each row to the best column.
In this example, we want to assign each swimming style to a student.

> This means that rows of the cost matrix show each swimming style and columns show each student.
> In other words, the upper table needs to be transposed for the cost matrix:

Reviewer: ```suggestion
Therefore, the table above needs to be transposed so that the rows correspond
with swimming styles and the columns correspond with students:
```

    >>> from scipy.optimize import linear_sum_assignment
    >>> row_ind, col_ind = linear_sum_assignment(cost)

The ``row_ind`` and ``col_ind`` are optimal assigned matrix indexes of the cost matrix:

::

    >>> row_ind
    array([0, 1, 2, 3])
    >>> col_ind
    array([0, 2, 3, 1])

The optimal assignment is:

::

    >>> styles = np.array(["backstroke", "breaststroke", "butterfly", "freestyle"])[row_ind]
    >>> students = np.array(["A", "B", "C", "D", "E"])[col_ind]
    >>> dict(zip(styles, students))
    {'backstroke': 'A', 'breaststroke': 'C', 'butterfly': 'D', 'freestyle': 'B'}

The optimal total medley time is:

::

    >>> cost[row_ind, col_ind].sum()
    163.89999999999998

Note that this result is not the same as the sum of the minimum times for each swimming style:

::

    >>> np.min(cost, axis=1).sum()
    161.39999999999998

> Because, student "C" is the best swimmer in both "breaststroke" and "butterfly" style.

Reviewer: ```suggestion
because student "C" is the best swimmer in both "breaststroke" and "butterfly" style.
```

> We need to decide either style is assigned to student "C" based on the total time.

Reviewer: We cannot assign student "C" to both styles, so we assigned student C to the _____ style and ____ to the ____ style to minimize the total time.

Author: Thanks. I fixed.
